### ISSUE TITLE
ToolRL is a multi-step Agent?

### ISSUE THREAD

**Issue by Zhaozeyong** [qiancheng0/ToolRL#40]
Hello, I’m researching ways to mitigate sparse-reward and reward-delay issues in agent training.
Could you tell me:

1.Approximately how many environment steps ToolRL typically needs to solve a task?
2.Does it provide a dense reward at every step (e.g., a format reward plus a correctness reward), or is the reward given only once—at the very end of execution?

**Comment by qiancheng0** [comment]
1. The envronment step typically varies from 1-5 in both training and evaluation and is not settled.
2. Yes, dense reward is provided as for each step a tool call is generally anticipated

**Comment by TongLiu-github** [comment]
In my understanding, the training data only contains single-turn tool-calling. 
What do you mean by "The envronment step typically varies from 1-5 in training"? Where is the more-than-one steps in training?

**Comment by qiancheng0** [comment]
Please refer to our paper (https://arxiv.org/pdf/2504.13958) Section 4.1 last paragraph: "Multi-step trajectories are decomposed into single-step instances, with prior dialogue history injected into the user prompt"

To simply put this, we decompose the multiturn trjectory offline into multiple single turn tool calling first. You may not see the original full trajectory because the data is shuffled. The environment step in the trajectory means originally the whole trajectory (before decomposition) usually contains 1-5 training steps. At the time we conduct our work we have not employed the currently common online multi-turn rollout paradigm.

**Comment by TongLiu-github** [comment]
Agree, good work thanks for reply.
